### ISSUE TITLE
RuntimeError: can't add a new key into hash during iteration

### ISSUE THREAD

**Issue by dlanderson** [scoutapp/scout_apm_ruby#205]
As reported by a user of `3.0.0.pre23`, there may be an iteration or store into a hash across threads not protected with synchronize:

```
RuntimeError: can't add a new key into hash during iteration
-
39 non-project frames
1 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/store.rb" line 9 in block in initialize
2 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/store.rb" line 18 in current_period
3 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/store.rb" line 26 in find_period
4 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/store.rb" line 34 in block in track!
5 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/store.rb" line 33 in synchronize
6 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/store.rb" line 33 in track!
7 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/layer_converters/metric_converter.rb" line 40 in record!
8 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/tracked_request.rb" line 326 in block in record!
9 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/tracked_request.rb" line 325 in each
10 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/tracked_request.rb" line 325 in inject
11 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/tracked_request.rb" line 325 in record!
12 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/synchronous_recorder.rb" line 27 in record!
13 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/tracked_request.rb" line 229 in stop_request
14 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/tracked_request.rb" line 123 in stop_layer
15 ile "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/tracer.rb" line 37 in instrument
16 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/tracer.rb" line 44 in instrument
17 File "/var/lib/gems/2.3.0/gems/scout_apm-3.0.0.pre23/lib/scout_apm/instruments/redis.rb" line 31 in call_with_scout_instruments
```

**Comment by cschneid** [comment]
It looks like the race is on two lines in store.rb:

```ruby
@reporting_periods = Hash.new { |h,k| h[k] = StoreReportingPeriod.new(k, @context) }
```

and 

```ruby
@reporting_periods.select { |time, rp| force || (time.timestamp < current_timestamp.timestamp) }.
  each   { |time, rp| write_reporting_period(layaway, time, rp) }
```

The call to `write_reporting_period` does File IO after a few more steps, so is relatively slow.  The Hash default value inserts a new ReportingPeriod when it is first requested for a certain time.

Likely the best way to solve this is to wrap both locations in the same Mutex used elsewhere in the Store class.

**Comment by cschneid** [comment]
Released in 2.4.16
